### PR TITLE
Blazor Web Assembly - Solution and Readme

### PR DESCRIPTION
**Pull request by EugeniyBurmistrov**

### CODE REVIEW

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 1]
```diff
@@ -0,0 +1,1114 @@
...
+  
+  > **NOTE**
+  >
+  > If you’ve used a pre-release version of our components or obtained a hotfix from DevExpress, NuGet packages will not be restored automatically (you will need to update them manually). For more information, please refer to the following article: [Updating Packages](https://docs.devexpress.com/GeneralInformation/118420/Installation/Install-DevExpress-Controls-Using-NuGet-Packages/Updating-Packages). Remember to enable the [Include prerelease](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages#installing-and-updating-pre-release-packages) option.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
  > If you’ve installed a pre-release version of our components or obtained a hotfix from DevExpress, NuGet packages will not be restored automatically (you will need to update them manually). For more information, please refer to the following article: [Updating Packages](https://docs.devexpress.com/GeneralInformation/118420/Installation/Install-DevExpress-Controls-Using-NuGet-Packages/Updating-Packages). Remember to enable the [Include prerelease](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages#installing-and-updating-pre-release-packages) option.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 2]
```diff
@@ -0,0 +1,1114 @@
...
+      public virtual string Content { get; set; }
+      public virtual ApplicationUser Author { get; set; }
+      public override void OnCreated() {
+         base.OnCreated();
+         Author = ObjectSpace.FindObject<ApplicationUser>(CriteriaOperator.Parse("ID=CurrentUserId()"));
+      }
+   }
+   ```
+
+   In the code sample above, `Post` class inherits [BaseObject](https://docs.devexpress.com/eXpressAppFramework/DevExpress.Persistent.BaseImpl.BaseObject) to simplify data model implementation. This tutorial makes use of the following `BaseObject` features:
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
   In the code sample above, the `Post` class inherits [BaseObject](https://docs.devexpress.com/eXpressAppFramework/DevExpress.Persistent.BaseImpl.BaseObject) to simplify data model implementation. This tutorial makes use of the following `BaseObject` features:
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 3]
```diff
@@ -0,0 +1,1114 @@
...
+      post.Content = "This is a FREE API for everybody";
+      post.Author=editorUser;
+      post = ObjectSpace.CreateObject<Post>();
+      post.Title = "Hello Blazor WebAssembly";
+      post.Content = "Please smash the like button to help our videos get discovered";
+      post.Author=editorUser;
+   }
+   ```
+
+### Configure Application Security
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]: The headings above (such as "### Declare a data model") only capitalize the first letter of the first word. Why do you capitalize "Application Security" here?

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 4]
```diff
@@ -0,0 +1,1114 @@
...
+   ```cmd
+   dotnet add package Blazored.LocalStorage --version 4.3.0
+   dotnet add package Microsoft.AspNetCore.Components.WebAssembly.Authentication --version 7.0.4
+   dotnet add package Microsoft.Extensions.Http --version 7.0.0
+   dotnet add package Microsoft.AspNetCore.Components.WebAssembly.Authentication --version 7.0.4
+   dotnet add package Simple.OData.Client --version 6.0.1
+   dotnet add package DevExpress.Blazor --version 22.2.4
+   ```
+
+### Register Installed Packages Services
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]: See my comment above regarding the capitalization of headings.

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 5]
```diff
@@ -0,0 +1,1114 @@
...
+   ```cs
+   @inject NavigationManager Navigation
+
+   @code {
+      protected override void OnInitialized() 
+         => Navigation.NavigateTo($"Login?returnUrl={Uri.EscapeDataString(Navigation.Uri)}");
+   }
+   ```
+
+6. Modify the `App.Razor` file. In the markup, replace the `RouteView` component with the `AuthorizedRouteView`.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
6. Modify the `App.Razor` file. In the markup, replace the `RouteView` component with `AuthorizedRouteView`.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 6]
```diff
@@ -0,0 +1,1114 @@
...
+   _App.Razor_:
+   ```razor
+   <AuthorizeRouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)">
+      <NotAuthorized>
+         <RedirectToLogin />
+      </NotAuthorized>
+   </AuthorizeRouteView>
+   ```
+
+7. To validate the authentication cookie, modify the `App.Razor` file as follows.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md:493]:
```suggestion
7. To validate the authentication cookie, modify the `App.Razor` file as follows:
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 7]
```diff
@@ -0,0 +1,1114 @@
...
+   Enclose all markup within `App.razor` with a `CascadingAuthenticationState` component.
+
+   _App.Razor_:
+   ```razor
+   <CascadingAuthenticationState>
+      @* App.Razor markup *@
+   </CascadingAuthenticationState>
+   ```
+
+   In the same file, handle the Router's `OnNavigateAsync` event to get the user profile or re-populate it if needed.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
   In the same file, handle the Router's `OnNavigateAsync` event to get the user profile or repopulate it if needed.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 8]
```diff
@@ -0,0 +1,1114 @@
...
+
+      private static IBoundClient<Post> ApplySorting(GridCustomDataSourceItemsOptions options, IBoundClient<Post> boundClient)
+         => options.SortInfo.Any() ? boundClient.OrderBy(options.SortInfo
+                  .Where(info => !info.DescendingSortOrder).Select(info => info.FieldName).ToArray())
+            .OrderByDescending(options.SortInfo
+                  .Where(info => info.DescendingSortOrder).Select(info => info.FieldName).ToArray()) : boundClient;
+   }
+   ```
+
+   For more information, refer to the following documentation topic: https://docs.devexpress.com/Blazor/DevExpress.Blazor.GridCustomDataSource
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
   For more information, refer to the following documentation topic: [GridCustomDataSource Class](https://docs.devexpress.com/Blazor/DevExpress.Blazor.GridCustomDataSource).
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 9]
```diff
@@ -0,0 +1,1114 @@
...
+   editorUser.Photo = ObjectSpace.CreateObject<MediaDataObject>();
+   editorUser.Photo.MediaData = GetResourceByName("Janete");
+   viewerUser.Photo = ObjectSpace.CreateObject<MediaDataObject>();
+   viewerUser.Photo.MediaData = GetResourceByName("John");
+   // ...
+   ```
+
+   > **NOTE**
+   >
+   > In our example code, the `GetResourceByName` method returns a byte array representation of an account image based on name. You can find an implementation for this method in the Web API project's [_DatabaseUpdate/Updater.cs_](./WebApi/DatabaseUpdate/Updater.cs) file. Note that this implementation requires the image resources to be compiled into the application's assembly (the .jpg files' `Build Action` option must be set to `Embedded resource`).
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
   > In our example code, the `GetResourceByName` method returns a byte array representation of an account image based on name. You can find an implementation for this method in the Web API project's [_DatabaseUpdate/Updater.cs_](./WebApi/DatabaseUpdate/Updater.cs) file. Note that this implementation requires image resources to be compiled into the application's assembly (the .jpg files' `Build Action` option must be set to `Embedded resource`).
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 10]
```diff
@@ -0,0 +1,1114 @@
...
+
+   Add a `CustomEndPointController` inside the `WebAPI/API` directory, and inject the `ISecurityProvider` and `IObjectSpaceFactory`. Implement a controller action that serves post author photos as follows.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
   Add `CustomEndPointController` inside the `WebAPI/API` directory, and inject `ISecurityProvider` and `IObjectSpaceFactory`. Implement a controller action that serves post author photos as follows.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 11]
```diff
@@ -0,0 +1,1114 @@
...
+   ```cs
+   [HttpGet(nameof(CanCreate))]
+   public IActionResult CanCreate(string typeName) {
+       var strategy = (SecurityStrategy)_securityProvider.GetSecurity();
+       var objectType = strategy.TypesInfo.PersistentTypes.First(info => info.Name  == typeName).Type;
+       return Ok(strategy.CanCreate(objectType));
+   }
+   ```
+
+2. In the Blazor WebAssembly project, update the `IWebAPI` service with a method with a method that queries the `CanCreate` endpoint to check if the current user can edit Posts.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
2. In the Blazor WebAssembly project, update the `IWebAPI` service with a method that queries the `CanCreate` endpoint to check if the current user can edit Posts.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 12]
```diff
@@ -0,0 +1,1114 @@
...
+               await _client.For<Post>().Set(new { instance.Title, instance.Content }).InsertEntryAsync();
+         }
+         else {
+               await _client.For<Post>().Key(instance.ID).Set(instance).UpdateEntryAsync();
+         }
+      }
+   }
+   ```
+
+4. Add a command column with _New_, _Edit_ and _Delete_ commands to the DxGrid and implement logic required process CRUD operations.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
4. Add a command column with _New_, _Edit_, and _Delete_ commands to the DxGrid and implement logic required to process CRUD operations.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md [hunk 13]
```diff
@@ -0,0 +1,1114 @@
...
+      // ...
+   }
+   public class WebAPI: IWebAPI {
+      public async Task ArchiveAsync(Post post) 
+         => await _httpClient.PostAsJsonAsync("CustomEndPoint/Archive", post);
+      // ...
+   }
+   ```
+
+3. In `FetchData.razor`, add a selection column to the DxGrid, a popup that displays the operation's result and a button that a user can click to archive the selected post.
```
Comment (albertov05) [inline EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]:
```suggestion
3. In `FetchData.razor`, add a selection column to DxGrid, a popup that displays the operation's result, and a button that a user clicks to archive the selected post.
```

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md (hunk 3 above)
Comment (EugeniyBurmistrov) [reply EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]: I fixed the capitalization

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md (hunk 4 above)
Comment (EugeniyBurmistrov) [reply EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]: Fixed

File: EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md (hunk 3 above)
Comment (albertov05) [reply EFCore/ASP.NetCore/Blazor.WebAssembly/Readme.md]: Thank you

**Pull request merged**